agent/check.go

```diff
@@ -128,19 +128,27 @@ func (c *CheckMonitor) check() {
 	}
 
 	// Wait for the check to complete
-	errCh := make(chan error, 2)
+	waitCh := make(chan error, 1)
 	go func() {
-		errCh <- cmd.Wait()
+		waitCh <- cmd.Wait()
 	}()
-	go func() {
-		if c.Timeout > 0 {
-			time.Sleep(c.Timeout)
+
+	timeout := 30 * time.Second
+	if c.Timeout > 0 {
+		timeout = c.Timeout
+	}
+	select {
+	case <-time.After(timeout):
+		if err := cmd.Process.Kill(); err != nil {
+			c.Logger.Printf("[WARN] Timed out running check '%s': error killing process: %v", c.Script, err)
 		} else {
-			time.Sleep(30 * time.Second)
+			c.Logger.Printf("[WARN] Timed out running check '%s'", c.Script)
 		}
-		errCh <- fmt.Errorf("Timed out running check '%s'", c.Script)
-	}()
-	err = <-errCh
+		err = <-waitCh
```

Comment:
We should set `err` to `fmt.Errorf("Timed out running check '%s'", c.Script)` to make sure the check gets set to critical (otherwise if the wait returns `nil` it can go passing, and then we just block on the `waitCh` but don't set `err`, so:

```
err = fmt.Errorf("Timed out running check '%s'", c.Script)
<-waitCh
```